File: Kernel Simulator/Misc/Probers/PlaceParse.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Public Module PlaceParse

    ''' <summary>
    ''' Probes the placeholders found in string
    ''' </summary>
    ''' <param name="text">Specified string</param>
    ''' <returns>A string that has the parsed placeholders</returns>
    Public Function ProbePlaces(text As String) As String

        EventManager.RaisePlaceholderParsing(text)
        Try
            Wdbg("I", "Parsing text for placeholders...")
            If text.Contains("<user>") Then
                Wdbg("I", "Username placeholder found.")
                text = text.Replace("<user>", signedinusrnm)
            End If
            If text.Contains("<ftpuser>") Then
                Wdbg("I", "FTP username placeholder found.")
                text = text.Replace("<ftpuser>", user)
            End If
            If text.Contains("<ftpaddr>") Then
                Wdbg("I", "FTP address placeholder found.")
                text = text.Replace("<ftpaddr>", ftpsite)
            End If
            If text.Contains("<currentftpdirectory>") Then
                Wdbg("I", "FTP directory placeholder found.")
                text = text.Replace("<currentftpdirectory>", currentremoteDir)
            End If
            If text.Contains("<currentftplocaldirectory>") Then
                Wdbg("I", "FTP local directory placeholder found.")
                text = text.Replace("<currentftplocaldirectory>", currDirect)
            End If
            If text.Contains("<sftpuser>") Then
                Wdbg("I", "SFTP username placeholder found.")
                text = text.Replace("<sftpuser>", SFTPUser)
            End If
            If text.Contains("<sftpaddr>") Then
                Wdbg("I", "SFTP address placeholder found.")
                text = text.Replace("<sftpaddr>", sftpsite)
            End If
            If text.Contains("<currentsftpdirectory>") Then
                Wdbg("I", "SFTP directory placeholder found.")
                text = text.Replace("<currentsftpdirectory>", SFTPCurrentRemoteDir)
            End If
            If text.Contains("<currentsftplocaldirectory>") Then
                Wdbg("I", "SFTP local directory placeholder found.")
                text = text.Replace("<currentsftplocaldirectory>", SFTPCurrDirect)
            End If
            If text.Contains("<mailuser>") Then
                Wdbg("I", "Mail username placeholder found.")
                text = text.Replace("<mailuser>", Mail_Authentication.UserName)
            End If
            If text.Contains("<mailaddr>") Then
                Wdbg("I", "Mail address placeholder found.")
                text = text.Replace("<mailaddr>", Mail_Authentication.Domain)
            End If
            If text.Contains("<currentmaildirectory>") Then
                Wdbg("I", "Mail directory placeholder found.")
                text = text.Replace("<currentmaildirectory>", IMAP_CurrentDirectory)
            End If
            If text.Contains("<host>") Then
                Wdbg("I", "Hostname placeholder found.")
                text = text.Replace("<host>", HName)
            End If
            If text.Contains("<currentdirectory>") Then
                Wdbg("I", "Current directory placeholder found.")
                text = text.Replace("<currentdirectory>", CurrDir)
            End If
            If text.Contains("<shortdate>") Then
                Wdbg("I", "Short Date placeholder found.")
                text = text.Replace("<shortdate>", KernelDateTime.ToShortDateString)
            End If
            If text.Contains("<longdate>") Then
                Wdbg("I", "Long Date placeholder found.")
                text = text.Replace("<longdate>", KernelDateTime.ToLongDateString)
            End If
            If text.Contains("<shorttime>") Then
                Wdbg("I", "Short Time placeholder found.")
                text = text.Replace("<shorttime>", KernelDateTime.ToShortTimeString)
            End If
            If text.Contains("<longtime>") Then
                Wdbg("I", "Long Time placeholder found.")
                text = text.Replace("<longtime>", KernelDateTime.ToShortDateString)
            End If
            If text.Contains("<date>") Then
                Wdbg("I", "Rendered Date placeholder found.")
                text = text.Replace("<date>", RenderDate)
            End If
            If text.Contains("<time>") Then
                Wdbg("I", "Rendered Time placeholder found.")
                text = text.Replace("<time>", RenderTime)
            End If
            If text.Contains("<timezone>") Then
                Wdbg("I", "Standard Time Zone placeholder found.")
                text = text.Replace("<timezone>", TimeZone.CurrentTimeZone.StandardName)
            End If
            If text.Contains("<summertimezone>") Then
                Wdbg("I", "Summer Time Zone placeholder found.")
                text = text.Replace("<summertimezone>", TimeZone.CurrentTimeZone.DaylightName)
            End If
            If text.Contains("<system>") Then
                Wdbg("I", "System placeholder found.")
                text = text.Replace("<system>", Environment.OSVersion.ToString)
            End If
            If text.Contains("<f:reset>") Then
                Wdbg("I", "Foreground color reset placeholder found.")
                text = text.Replace("<f:reset>", New Color(NeutralTextColor).VTSequenceForeground)
            End If
            If text.Contains("<b:reset>") Then
                Wdbg("I", "Background color reset placeholder found.")
                text = text.Replace("<b:reset>", New Color(BackgroundColor).VTSequenceBackground)
            End If
            If text.Contains("<f:") Then
                Wdbg("I", "Foreground color placeholder found.")
                Do While text.Contains("<f:")
                    Dim SequenceSubstring As String = text.Substring(text.IndexOf("<f:"), Finish:=text.IndexOf(">"))
                    Dim PlainSequence As String = SequenceSubstring.Substring(3, SequenceSubstring.Length - 1 - 3)
                    Dim VTSequence As String = New Color(PlainSequence).VTSequenceForeground
                    text = text.Replace(SequenceSubstring, VTSequence)
                Loop
            End If
            If text.Contains("<b:") Then
                Wdbg("I", "Background color placeholder found.")
                Do While text.Contains("<b:")
                    Dim SequenceSubstring As String = text.Substring(text.IndexOf("<b:"), Finish:=text.IndexOf(">"))
                    Dim PlainSequence As String = SequenceSubstring.Substring(3, SequenceSubstring.Length - 1 - 3)
                    Dim VTSequence As String = New Color(PlainSequence).VTSequenceBackground
                    text = text.Replace(SequenceSubstring, VTSequence)
                Loop
            End If
            If text.Contains("<$") Then
                Wdbg("I", "UESH variable placeholder found.")
                Do While text.Contains("<$")
                    Dim ShellVariableSubstring As String = text.Substring(text.IndexOf("<$"), Finish:=text.IndexOf(">"))
                    Dim PlainShellVariable As String = ShellVariableSubstring.Substring(1, ShellVariableSubstring.Length - 1 - 1)
                    text = text.Replace(ShellVariableSubstring, GetVariable(PlainShellVariable))
                Loop
            End If
            EventManager.RaisePlaceholderParsed(text)
        Catch ex As Exception
            WStkTrc(ex)
            If DebugMode = True Then
                Write(DoTranslation("Error trying to parse placeholders. {0} - Stack trace:") + vbNewLine + ex.StackTrace, True, ColTypes.Error, ex.Message)
            Else
                Write(DoTranslation("Error trying to parse placeholders. {0}"), True, ColTypes.Error, ex.Message)
            End If
        End Try
        Return text

    End Function

End Module

File: Kernel Simulator/Misc/Probers/PlaceParse.cs
//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System.IO;
using KS.ConsoleBase.Colors;
using KS.Files.Folders;
using KS.Languages;
using KS.Login;
using KS.Misc.Writers.DebugWriters;
using KS.Network.FTP;
using KS.Network.Mail;
using KS.Network.SFTP;
using KS.Scripting;
using KS.TimeDate;
using Terminaux.Colors;

namespace KS.Misc.Probers
{
    public static class PlaceParse
    {

        /// <summary>
        /// Probes the placeholders found in string
        /// </summary>
        /// <param name="text">Specified string</param>
        /// <param name="ThrowIfFailure">Throw if the placeholder parsing fails</param>
        /// <returns>A string that has the parsed placeholders</returns>
        public static string ProbePlaces(string text, bool ThrowIfFailure = false)
        {
            Kernel.Kernel.KernelEventManager.RaisePlaceholderParsing(text);
            try
            {
                // Parse the text for the following placeholders:
                DebugWriter.Wdbg(DebugLevel.I, "Parsing text for placeholders...");

                // -> User placeholder
                if (text.Contains("<user>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Username placeholder found.");
                    text = text.Replace("<user>", Login.Login.CurrentUser.Username);
                }

                // -> FTP user placeholder
                if (text.Contains("<ftpuser>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "FTP username placeholder found.");
                    text = text.Replace("<ftpuser>", FTPShellCommon.FtpUser);
                }

                // -> FTP address placeholder
                if (text.Contains("<ftpaddr>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "FTP address placeholder found.");
                    text = text.Replace("<ftpaddr>", FTPShellCommon.FtpSite);
                }

                // -> Current FTP directory placeholder
                if (text.Contains("<currentftpdirectory>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "FTP directory placeholder found.");
                    text = text.Replace("<currentftpdirectory>", FTPShellCommon.FtpCurrentRemoteDir);
                }

                // -> Current FTP local directory placeholder
                if (text.Contains("<currentftplocaldirectory>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "FTP local directory placeholder found.");
                    text = text.Replace("<currentftplocaldirectory>", FTPShellCommon.FtpCurrentDirectory);
                }

                // -> Current FTP local directory name placeholder
                if (text.Contains("<currentftplocaldirectoryname>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "FTP local directory name placeholder found.");
                    text = text.Replace("<currentftplocaldirectoryname>", new DirectoryInfo(FTPShellCommon.FtpCurrentDirectory).Name);
                }

                // -> SFTP user placeholder
                if (text.Contains("<sftpuser>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "SFTP username placeholder found.");
                    text = text.Replace("<sftpuser>", SFTPShellCommon.SFTPUser);
                }

                // -> SFTP address placeholder
                if (text.Contains("<sftpaddr>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "SFTP address placeholder found.");
                    text = text.Replace("<sftpaddr>", SFTPShellCommon.SFTPSite);
                }

                // -> Current SFTP directory placeholder
                if (text.Contains("<currentsftpdirectory>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "SFTP directory placeholder found.");
                    text = text.Replace("<currentsftpdirectory>", SFTPShellCommon.SFTPCurrentRemoteDir);
                }

                // -> Current SFTP local directory placeholder
                if (text.Contains("<currentsftplocaldirectory>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "SFTP local directory placeholder found.");
                    text = text.Replace("<currentsftplocaldirectory>", SFTPShellCommon.SFTPCurrDirect);
                }

                // -> Current SFTP local directory name placeholder
                if (text.Contains("<currentsftplocaldirectoryname>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "SFTP local directory name placeholder found.");
                    text = text.Replace("<currentsftplocaldirectoryname>", new DirectoryInfo(SFTPShellCommon.SFTPCurrDirect).Name);
                }

                // -> Mail user placeholder
                if (text.Contains("<mailuser>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Mail username placeholder found.");
                    text = text.Replace("<mailuser>", MailLogin.Mail_Authentication.UserName);
                }

                // -> Mail address placeholder
                if (text.Contains("<mailaddr>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Mail address placeholder found.");
                    text = text.Replace("<mailaddr>", MailLogin.Mail_Authentication.Domain);
                }

                // -> Current mail directory placeholder
                if (text.Contains("<currentmaildirectory>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Mail directory placeholder found.");
                    text = text.Replace("<currentmaildirectory>", MailShellCommon.IMAP_CurrentDirectory);
                }

                // -> Hostname placeholder
                if (text.Contains("<host>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Hostname placeholder found.");
                    text = text.Replace("<host>", Kernel.Kernel.HostName);
                }

                // -> Current directory placeholder
                if (text.Contains("<currentdirectory>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Current directory placeholder found.");
                    text = text.Replace("<currentdirectory>", CurrentDirectory.CurrentDir);
                }

                // -> Current directory name placeholder
                if (text.Contains("<currentdirectoryname>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Current directory name placeholder found.");
                    text = text.Replace("<currentdirectoryname>", new DirectoryInfo(CurrentDirectory.CurrentDir).Name);
                }

                // -> Short date placeholder
                if (text.Contains("<shortdate>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Short Date placeholder found.");
                    text = text.Replace("<shortdate>", TimeDate.TimeDate.KernelDateTime.ToShortDateString());
                }

                // -> Long date placeholder
                if (text.Contains("<longdate>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Long Date placeholder found.");
                    text = text.Replace("<longdate>", TimeDate.TimeDate.KernelDateTime.ToLongDateString());
                }

                // -> Short time placeholder
                if (text.Contains("<shorttime>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Short Time placeholder found.");
                    text = text.Replace("<shorttime>", TimeDate.TimeDate.KernelDateTime.ToShortTimeString());
                }

                // -> Long time placeholder
                if (text.Contains("<longtime>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Long Time placeholder found.");
                    text = text.Replace("<longtime>", TimeDate.TimeDate.KernelDateTime.ToShortDateString());
                }

                // -> Date placeholder
                if (text.Contains("<date>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Rendered Date placeholder found.");
                    text = text.Replace("<date>", TimeDateRenderers.RenderDate());
                }

                // -> Time placeholder
                if (text.Contains("<time>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Rendered Time placeholder found.");
                    text = text.Replace("<time>", TimeDateRenderers.RenderTime());
                }

                // -> Timezone placeholder
                if (text.Contains("<timezone>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Standard Time Zone placeholder found.");
                    text = text.Replace("<timezone>", TimeZoneInfo.Local.StandardName);
                }

                // -> Summer timezone placeholder
                if (text.Contains("<summertimezone>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Summer Time Zone placeholder found.");
                    text = text.Replace("<summertimezone>", TimeZoneInfo.Local.DaylightName);
                }

                // -> System placeholder
                if (text.Contains("<system>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "System placeholder found.");
                    text = text.Replace("<system>", Environment.OSVersion.ToString());
                }

                // -> Newline placeholder
                if (text.Contains("<newline>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Newline placeholder found.");
                    text = text.Replace("<newline>", Kernel.Kernel.NewLine);
                }

                // -> User dollar sign placeholder
                if (text.Contains("<dollar>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Dollar placeholder found.");
                    text = text.Replace("<dollar>", UserManagement.GetUserDollarSign());
                }

                // -> Foreground color reset placeholder
                if (text.Contains("<f:reset>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Foreground color reset placeholder found.");
                    text = text.Replace("<f:reset>", KernelColorTools.NeutralTextColor.VTSequenceForeground);
                }

                // -> Background color reset placeholder
                if (text.Contains("<b:reset>"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Background color reset placeholder found.");
                    text = text.Replace("<b:reset>", KernelColorTools.BackgroundColor.VTSequenceBackground);
                }

                // -> Foreground color placeholder
                if (text.Contains("<f:"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Foreground color placeholder found.");
                    while (text.Contains("<f:"))
                    {
                        int StartForegroundIndex = text.IndexOf("<f:");
                        int EndForegroundIndex = text.Substring(text.IndexOf("<f:")).IndexOf(">");
                        string SequenceSubstring = text.Substring(text.IndexOf("<f:"), length: EndForegroundIndex + 1);
                        string PlainSequence = SequenceSubstring.Substring(3, SequenceSubstring.Length - 1 - 3);
                        string VTSequence = new Color(PlainSequence).VTSequenceForeground;
                        text = text.Replace(SequenceSubstring, VTSequence);
                    }
                }

                // -> Background color placeholder
                if (text.Contains("<b:"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "Background color placeholder found.");
                    while (text.Contains("<b:"))
                    {
                        int StartBackgroundIndex = text.IndexOf("<b:");
                        int EndBackgroundIndex = text.Substring(text.IndexOf("<b:")).IndexOf(">");
                        string SequenceSubstring = text.Substring(text.IndexOf("<b:"), length: EndBackgroundIndex + 1);
                        string PlainSequence = SequenceSubstring.Substring(3, SequenceSubstring.Length - 1 - 3);
                        string VTSequence = new Color(PlainSequence).VTSequenceBackground;
                        text = text.Replace(SequenceSubstring, VTSequence);
                    }
                }

                // -> UESH variable placeholder
                if (text.Contains("<$"))
                {
                    DebugWriter.Wdbg(DebugLevel.I, "UESH variable placeholder found.");
                    while (text.Contains("<$"))
                    {
                        int StartShellVariableIndex = text.IndexOf("<$");
                        int EndShellVariableIndex = text.Substring(text.IndexOf("<$")).IndexOf(">");
                        string ShellVariableSubstring = text.Substring(text.IndexOf("<$"), length: EndShellVariableIndex + 1);
                        string PlainShellVariable = ShellVariableSubstring.Substring(1, ShellVariableSubstring.Length - 1 - 1);
                        text = text.Replace(ShellVariableSubstring, UESHVariables.GetVariable(PlainShellVariable));
                    }
                }

                // If successful, raise the parsed event
                Kernel.Kernel.KernelEventManager.RaisePlaceholderParsed(text);
            }
            catch (Exception ex)
            {
                DebugWriter.Wdbg(DebugLevel.E, "Failed to parse placeholder {0}: {1}", text, ex.Message);
                DebugWriter.WStkTrc(ex);
                Kernel.Kernel.KernelEventManager.RaisePlaceholderParseError(text, ex);
                if (ThrowIfFailure)
                    throw new Kernel.Exceptions.InvalidPlaceholderException(Translate.DoTranslation("Error trying to parse placeholders. {0}"), ex.Message);
            }
            return text;
        }

    }
}